src/bevy_compat/keyboard.rs

```diff
@@ -0,0 +1,647 @@
+    prelude::*,
+    window::PrimaryWindow,
+};
+use crossterm::event::KeyModifiers;
+
+use crate::{
+    event::{InputSet, KeyEvent},
+    kitty::KittyEnabled,
+};
+
+/// Pass crossterm key events through to bevy's input system.
+///
+/// You can use bevy's regular `ButtonInput<KeyCode>` or bevy_ratatui's
+/// `EventReader<KeyEvent>`.
+pub struct KeyboardPlugin;
+
+impl Plugin for KeyboardPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_plugins(bevy::input::InputPlugin)
+            .add_systems(Startup, setup_window)
+            .add_systems(PreUpdate, send_key_events.in_set(InputSet::EmitBevy));
+    }
+}
+
+#[derive(Debug, Deref, DerefMut)]
+struct Modifiers(KeyModifiers);
+
+impl Default for Modifiers {
+    fn default() -> Self {
+        Self(KeyModifiers::empty())
+    }
+}
+
+// I wish KeyboardInput had the Hash derive.
+// TODO: Drop this if KeyboardInput get a Hash impl.
+#[derive(Deref, DerefMut, PartialEq, Eq)]
+struct KeyInput(KeyboardInput);
+
+impl Hash for KeyInput {
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: Hasher,
+    {
+        self.key_code.hash(state);
+        self.logical_key.hash(state);
+        self.state.hash(state);
+        self.window.hash(state);
+    }
+}
+
+/// This is a double-buffer for the last keys pressed.
+///
+/// - `.0` is current frame's last keys pressed.
+/// - `.1` is the next frame's last keys pressed.
+#[derive(Default)]
+struct LastPress(HashSet<KeyInput>, HashSet<KeyInput>);
+
+impl LastPress {
+    fn swap(&mut self) {
+        std::mem::swap(&mut self.0, &mut self.1);
+    }
+}
+
+fn send_key_events(
+    mut keys: EventReader<KeyEvent>,
+    kitty_enabled: Option<Res<KittyEnabled>>,
+    window: Query<Entity, With<PrimaryWindow>>,
+    mut modifiers: Local<Modifiers>,
+    mut last_pressed: Local<LastPress>,
+    mut keyboard_input: EventWriter<KeyboardInput>,
+) {
+    let bevy_window = window.single();
+    for key_event in keys.read() {
+        if let Some((bevy_event, mods)) = key_event_to_bevy(key_event, bevy_window) {
+            // dbg!(mods, *modifiers);
+            if mods != **modifiers {
+                let delta = mods.symmetric_difference(**modifiers);
+                for flag in delta {
+                    let state = if mods.contains(flag) {
+                        // This flag has been added.
+                        bevy::input::ButtonState::Pressed
+                    } else {
+                        // modifiers.contains(flag)
+                        // This flag has been removed.
+                        bevy::input::ButtonState::Released
+                    };
+                    keyboard_input.send(modifier_to_bevy(
+                        crossterm_modifier_to_bevy_key(flag),
+                        state,
+                        bevy_window,
+                    ));
+                }
+                **modifiers = mods;
+            }
+            if kitty_enabled.is_none() {
```

Comment:
Just asking for clarification here, is the aim of this code just to keep bevy's bookkeeping of what is "currently pressed" accurate? In my own testing, pressing and holding a key results in a rapid series of `Repeat` and `Release` events as repeat events are not sent at the same cadence as the framerate (and there is a half second delay before repeat events are sent at all), but I suppose that it makes sure a release event is always sent to clean up bevy's registry.

I'm not sure how best to solve this, but the reason I think this may be slightly problematic is that when terminals report "kitty support" it doesn't seem to mean they support every feature (for example Iterm2 will say that kitty support is enabled but then will not report key release events). Therefore when implementing fallbacks, you can't really look at the `KittyEnabled` flag, you have to conservatively stick to a non-kitty mode until you detect the feature you need (for example when you actually see a release event).

I don't know whether it makes sense to hold up this PR for it (opinion, @joshka?), but it should be noted that there is now a footgun for users looking at the normal bevy events where the accuracy of bevy's `pressed()` is very terminal-specific and somewhat un-obviously you have to listen for the original crossterm release events in order to actually determine support.

Reply:
Regarding `seen_key_released`, I would say that yes, it should be surfaced to the user, as there are generally going to be gameplay-specific fallbacks that need to be implemented. The example I would give is moving a character— if release events are present, just moving the character while an arrow is pressed is fine, but if release events are not present, meaning there is a delay before our fake release events are sent, the dev might want to add some smooth momentum falloff so that the delay feels like "taking a step". Might be cleanest to make `KittyEnabled` a struct that contains flags for specific feature support that bevy devs might be interested on.

Regarding this specific fix, I don't think its quite ready yet. Testing across Terminal.app/Alacritty/iTerm, there are a few broken behaviors. With kitty support, holding a modifier and then a pressing a number key (for e.g. '#' or '@') a second later results in just the number being registered, as a Release event is sent for the shift key before the second key is pressed and before the shift key is raised. Without kitty support, it appears possible to still flip the `seen_key_released` flag by pressing modifier keys, which subsequently makes it possible to build up a long list of "currently pressed" keys as they stop getting removed. Also in the current code I don't believe regular release events flip that flag, so the fallback continues to apply in Alacritty, a terminal with full kitty support. In general, papering over the lack of release events by listening for repeat events over delay is not the most reliable, as the delay before a key sends a Repeat event is an OS-level user configurable value, so even if we let devs configure that `ReleaseKey` resource, there's not really a value that will work for all players.

Zooming out, if there is a clever way to solve these problems and keep the registry of keys clear then all the better! But prioritizing predictability for the user over full compatibility, rather than getting too fancy I think we should just meet the following:

1. The second any true `crossterm::event::KeyEventKind::Release` event is detected, the simplest version of forwarding should occur. Alacritty/Kitty correctly report press and release events for even modifier keys on their own, so we can just pass this along to bevy.
2. We should keep the translation layer as simple as possible and clearly document that devs should listen for keypress events, and not rely on the `pressed()` methods, as there isn't a reliable cross-platform way of ensuring its accuracy.

Very happy to be pushed back on here (you have far more expertise with crossterm than I do) but I am worried that the second-delay timing approach is not going to be flexible across platforms and gameplay styles.

Reply:
Thank you for this fulsome reply. I have surfaced the capability detection to the user and provided a means for them to select what to emulate. By default it emulates whatever hasn't been detected.

I hear what you're saying about having a simple pass thru system. I hope that `send_key_events_no_emulation()` meets that criteria.

Regarding emulation, I did find that instead of emulation being on or off, there are at least four cases. WezTerm provides key releases but not modifier events. So we may not get key releases and we may not get modifier events.

The way I've structured it is, `send_key_events_with_emulation()` is called. If we're not emulating anything, it merely delegates to the no emulation function.

Kick the tires and see what other bugs you can get to jump out.

Reply:
After the latest commits I gave it another go and it is much more robust! I did find a couple of issues though:

1. For wezterm/kitty/alacritty, the first keypress after starting an app will register a press and release event, adding and then removing the key from bevy's list of pressed keys, and correctly switching into full support mode. However, a second later a fake event is sometimes still triggered which actually _adds_ the pressed key back into bevy's list where it stays permanently. It is not consistent, there is some kind of finicky frame-timing condition that causes this.

2. For kitty/alacritty, if you press and hold a key, repeat events are not handled correctly. It appears (take with a grain of salt, this is just looking at the output of the example) that the first repeat event causes a fake release event to be sent, removing the key from bevy's list, and then subsequent repeat events cause it to be re-added to bevy's list. Also, the re-added instance of the key is not removed when the key is released, polluting the list. I believe the only reason that this issue doesn't come up in wezterm is that wezterm doesn't send repeat events, as far as I can tell.

Other than these two, the emulation seems solid and worked in the five terminals I tested, and worked in a couple of `bevy_ratatui_render` examples as well!

Reply:
I'm going to merge this PR with these bugs and will make a few changes mentioned [below](https://github.com/joshka/bevy_ratatui/pull/13#pullrequestreview-2174340840).

I guessed there might be the race condition mentioned in 1 when reading the code, so it's good you saw this IRL. I also was curious about how the repeat events were handled.